sky/dag.py

```diff
@@ -1,75 +1,253 @@
 
+        Args:
+            name: Optional name for the DAG.
+        """
+        self.name = name
+        self._task_name_lookup: Dict[str, 'task.Task'] = {}
+        self.edges: Dict['task.Task', Set['task.Task']] = {}
         self.graph = nx.DiGraph()
-        self.name: Optional[str] = None
+
+    @property
+    def tasks(self) -> List['task.Task']:
+        """Return a list of all tasks in the DAG."""
+        return list(self._task_name_lookup.values())
+
+    def _get_task(self, task_or_name: TaskOrName) -> 'task.Task':
+        """Get a task object from a task or its name.
+
+        Args:
+            task_or_name: Either a Task object or the name of a task.
+
+        Returns:
+            The Task object.
+
+        Raises:
+            ValueError: If the task name is not found in the DAG.
+        """
+        if not isinstance(task_or_name, str):
+            return task_or_name
+        name = task_or_name
+        if name not in self._task_name_lookup:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(f'Task {name} not found in DAG')
+        return self._task_name_lookup[name]
 
     def add(self, task: 'task.Task') -> None:
+        """Add a task to the DAG.
+
+        Args:
+            task: The Task object to add.
+
+        Raises:
+            ValueError: If the task already exists in the DAG or if its name
+            is already used.
+        """
+        if task.name is None:
+            task.name = common_utils.get_unique_task_name(task)
+        if task.name in self._task_name_lookup:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'Task {task.name!r} already exists in the DAG.'
+                    f' Or the task name is already used by another task.')
         self.graph.add_node(task)
-        self.tasks.append(task)
+        self._task_name_lookup[task.name] = task
+
+    def remove(self, task: Union['task.Task', str]) -> None:
+        """Remove a task from the DAG.
+
+        Args:
+            task: The Task object or name of the task to remove.
+
+        Raises:
+            ValueError: If the task is still being depended on by other tasks.
+        """
+        task = self._get_task(task)
 
-    def remove(self, task: 'task.Task') -> None:
-        self.tasks.remove(task)
+        # TODO(andy): Stuck by optimizer's wrong way to remove dummy sources
+        # and sink nodes.
+        # if dependents:
+        #     dependent_names = ', '.join([dep.name for dep in dependents])
+        #     with ux_utils.print_exception_no_traceback():
+        #         raise ValueError(f'Task {task.name} is still being depended '
+        #                          f'by tasks {dependent_names!r}. Try to '
+        #                          'remove the dependencies first.')
+        # Here's a workaround, proactively remove all downstream edges.
+        dependents = self.get_downstream(task)
```

Comment:
```suggestion
        downstreams = self.get_downstream(task)
```
pls check naming in other places as well ;)